iroh-relay/src/server/clients.rs

```diff
@@ -1,228 +1,129 @@
+pub(super) struct Clients(Arc<Inner>);
+
 #[derive(Debug, Default)]
-pub(super) struct Clients {
+struct Inner {
     /// The list of all currently connected clients.
-    inner: HashMap<NodeId, Client>,
-    /// The next connection number to use.
-    conn_num: usize,
+    clients: DashMap<NodeId, Client>,
+    /// Map of which client has sent where
+    sent_to: DashMap<NodeId, HashSet<NodeId>>,
 }
 
 impl Clients {
-    pub async fn shutdown(&mut self) {
-        trace!("shutting down {} clients", self.inner.len());
+    pub async fn shutdown(&self) {
+        let keys: Vec<_> = self.0.clients.iter().map(|x| *x.key()).collect();
+        trace!("shutting down {} clients", keys.len());
+        let clients = keys.into_iter().filter_map(|k| self.0.clients.remove(&k));
 
         futures_buffered::join_all(
-            self.inner
-                .drain()
-                .map(|(_, client)| async move { client.shutdown().await }),
+            clients.map(|(_, client)| async move { client.shutdown().await }),
         )
         .await;
     }
 
-    /// Record that `src` sent or forwarded a packet to `dst`
-    pub fn record_send(&mut self, src: &NodeId, dst: NodeId) {
-        if let Some(client) = self.inner.get_mut(src) {
-            client.record_send(dst);
-        }
-    }
-
-    pub fn contains_key(&self, key: &NodeId) -> bool {
-        self.inner.contains_key(key)
-    }
-
-    pub fn has_client(&self, key: &NodeId, conn_num: usize) -> bool {
-        if let Some(client) = self.inner.get(key) {
-            return client.conn.conn_num == conn_num;
-        }
-        false
-    }
-
-    fn next_conn_num(&mut self) -> usize {
-        let conn_num = self.conn_num;
-        self.conn_num = self.conn_num.wrapping_add(1);
-        conn_num
-    }
-
     /// Builds the client handler and starts the read & write loops for the connection.
-    pub async fn register(&mut self, client_config: ClientConnConfig) {
-        let key = client_config.node_id;
-        trace!("registering client: {:?}", key);
-        let conn_num = self.next_conn_num();
-        let client = ClientConn::new(client_config, conn_num);
-        // TODO: in future, do not remove clients that share a NodeId, instead,
-        // expand the `Client` struct to handle multiple connections & a policy for
-        // how to handle who we write to when multiple connections exist.
-        let client = Client::new(client);
-        if let Some(old_client) = self.inner.insert(key, client) {
-            warn!("multiple connections found for {key:?}, pruning old connection",);
+    pub async fn register(&self, client_config: Config) {
+        let node_id = client_config.node_id;
+        trace!(node_id = node_id.fmt_short(), "registering client");
+
+        let client = Client::new(client_config, self);
+        if let Some(old_client) = self.0.clients.insert(node_id, client) {
+            warn!("multiple connections found for {node_id:?}, pruning old connection",);
```

Comment:
```suggestion
            warn!(node_id = node_id.fmt_short(), "multiple connections found for node, pruning old connection",);
```
This will make deduplicating this message on the ops side more easily.

Reply:
done